design-documents/promises.md

```diff
@@ -0,0 +1,240 @@
+* _wait_ method does not throw an exception if the promise is rejected
+nor does it return the result of the resolved promise ([see explanation](#wait-not-unwrapping-promises))
+* If an exception occurs during an asynchronous operation and no _otherwise_ callback is
+provided then it will just be rethrown
+### API
+##### Promise to be used by client code
+When client code receives a promise from calling another object's method
+it shouldn't have access to _resolve_ and _reject_ methods, it should only be able to
+provide callbacks to process promised results and to wait for promised operation's execution.
+ 
+This interface will be used as the return type of methods returning promises.
+```php
+interface PromiseInterface
+{
+    /**
+     * @throws PromiseProcessedException When callback was alredy provided.
+     */
+    public function then(callable $callback): void;
+    
+    /**
+     * @throws PromiseProcessedException When callback was alredy provided.
+     */
+    public function otherwise(callable $callback): void;
+    
+    public function wait(): void;
+}
+```
+##### Promise to be created
+This promise will be created by asynchronous code
+```php
+interface ResultPromiseInterface extends PromiseInterface
+{
+    public function resolve($value): void;
+    
+    public function reject(\Throwable $exception): void;
+}
+```
+
+### Implementation
+A wrapper around [Guzzle Promises](https://github.com/guzzle/promises) will be created to implement the APIs above. Guzzle Promises fit
+most important criteria - they allow synchronous execution as well as asynchronous. It's a mature
+and well-known library and, while we would have to add guzzle/promises to our composer.json,
+the library is already required in Magento indirectly - we won't be actually adding a new dependency.
+ 
+There are other libraries like [Reactphp Promises](https://github.com/reactphp/promise) but they either do not provide synchronous way
+to interact with promises or are not as refined.
+ 
+### Explanations
+##### Forwarding
+Consider this code
+```php
+$promise = $this->anotherObject->doStuff();
+$promise->then($doStuffCallback)
+    ->otherwise($processErrorCallback)
+    ->otherwise($processAnotherError)
+    ->then($doOtherStuff)
+    ->otherwise($processErrorCallback);
+```
+Does 1st _then_ return a forwarded promise? Or is it the same object?
+Then what promise is the second _otherwise_ callback for?
+Code looking like this is confusing and it will be much cleaner if we don't use forwarding.
+```php
+$doStuffOperation = $this->someObject->doStuff();
+$result = null;
+$doStuffOperation->then(function ($response) use (&$result) { $result = $response; });
+$simultaneousOperation = $this->otherObject->processStuff();
+$doStuffOperation->wait();
+$updated = null;
+$saveOperation = $this->repo->save($result);
+$saveOperation->then(function ($result) use (&$updated) { $updated = $result; });
+
+//Waiting for all
+$saveOperation->wait();
```

Comment:
Should we call `wait` for both async and sync approaches?

Reply:
you would call 'wait' to wait for an operation to complete if it's required, you can just provide a callback to 'then' method if you don't need waiting